### PR TITLE
Upgrade to boost 1 83

### PR DESCRIPTION
**Pull request by edsavage**
Upgrade to Boost 1.83 libraries

We have been using Boost 1.77 for approximately 2 years now, and it's good practice to upgrade occasionally.
    
As Boost 1.83 has just been released it makes sense to attempt to upgrade to it.

Labelling as a `>non-issue` as this change should have no outwardly facing effects.

### CODE REVIEW

File: build-setup/windows.md
Comment (droberts195) [inline build-setup/windows.md:142]: Why is this substitution no longer used? Did the Boost maintainers decide to start off with a smaller hash table in the official code? This optimisation was originally included because the anomaly detection code creates large numbers of hash maps with one item in them, and reserving 17 slots in such cases noticeably increases the memory usage of the `autodetect` process for high cardinality jobs.

Comment (edsavage) [reply build-setup/windows.md:142]: The `BOOST_UNORDERED_PRIMES` definition is no longer present after `Boost 1.79`, so there's no longer anything to substitute.

Comment (droberts195) [reply build-setup/windows.md:142]: Please can you grep the unordered code and check if another file has a similar set of primes in it. If you grep for 17 and 29 on the same line that might find it. I’m a bit worried we’ll get reports of greatly increased model memory sizes if the memory usage of a size 1 hash map dramatically increases.

Comment (droberts195) [reply build-setup/windows.md:142]: I suggest adding `(3ul)` to the beginning of that line and see if everything still works and memory usage is reduced.

File: dev-tools/download_windows_deps.ps1
Comment (droberts195) [inline dev-tools/download_windows_deps.ps1:14]: The condition here needs to be changed to look for a file that only exists in Boost 1.83.

https://github.com/elastic/ml-cpp/pull/638/files#diff-0a356335cbaadf7c9897e08805d11e4d124df92f6b166867982739296cd008b1 is an example from a previous upgrade.

(It's much easier to detect if latest Boost is present than to detect latest PyTorch.)

File: build-setup/macos.md
Comment (droberts195) [inline build-setup/macos.md:90]:
```suggestion
Edit `boost/unordered/detail/prime_fmod.hpp` and change line 134 from
```

File: cmake/variables.cmake
Comment (droberts195) [inline cmake/variables.cmake:229]:
```suggestion
find_package(Boost 1.83.0 EXACT REQUIRED COMPONENTS iostreams filesystem program_options regex date_time log log_setup thread unit_test_framework json)
```

File: build-setup/linux.md
Comment (droberts195) [inline build-setup/linux.md:192]:
```suggestion
Edit `boost/unordered/detail/prime_fmod.hpp` and change line 134 from:
```

File: build-setup/windows.md
Comment (droberts195) [inline build-setup/windows.md:133]:
```suggestion
Edit `boost/unordered/detail/prime_fmod.hpp` and change line 134 from:
```

File: build-setup/linux.md
Comment (droberts195) [inline build-setup/linux.md:208]: The new flag `__STDC_FORMAT_MACROS` isn't being used with the `linux-aarch64` Docker image. I don't see why we'd need the flag on `x86_64` but not `aarch64` when we're using the same compiler and standard library version on both.

File: dev-tools/docker/build_macosx_build_image.sh
Comment (droberts195) [inline dev-tools/docker/build_macosx_build_image.sh:14]:
```suggestion
# C++ code for Intel macOS
```

File: dev-tools/docker/linux_aarch64_native_image/Dockerfile
Comment (droberts195) [inline dev-tools/docker/linux_aarch64_native_image/Dockerfile:83]: These are the lines where I'd expect to see `cflags='-D__STDC_FORMAT_MACROS'` if it's needed at all.

File: lib/maths/analytics/unittest/CBoostedTreeTest.cc
Comment (droberts195) [inline lib/maths/analytics/unittest/CBoostedTreeTest.cc:3089]: Instead of adding the uplift here it should be added in https://github.com/elastic/ml-cpp/blob/c28c82c0c29e13990d169fbea5a6e1836e3a1b91/include/core/CMemoryDef.h#L208-L210 https://github.com/elastic/ml-cpp/blob/c28c82c0c29e13990d169fbea5a6e1836e3a1b91/include/core/CMemoryDef.h#L220-L221  https://github.com/elastic/ml-cpp/blob/c28c82c0c29e13990d169fbea5a6e1836e3a1b91/include/core/CMemoryDef.h#L492-L493 and https://github.com/elastic/ml-cpp/blob/c28c82c0c29e13990d169fbea5a6e1836e3a1b91/include/core/CMemoryDef.h#L529-L530

The new memory layout is detailed here: https://www.boost.org/doc/libs/1_83_0/libs/unordered/doc/html/unordered.html#structures_closed_addressing_containers

The formulas should be updated to match this. As far as I can see the key part is:

> there’s only 4 bits of space overhead per bucket introduced by the bucket groups.

So we should add on an extra half a byte per bucket (rounded up to the next higher multiple of `sizeof(std::size_t))`) to account for these bucket groups.

Please give this a try and see if it allows the 1% hack here to be removed. If it does then please also add a link to that section of the Boost docs as part of a comment to explain why we're adding this strange amount.

File: dev-tools/docker/linux_aarch64_native_image/Dockerfile
Comment (edsavage) [reply dev-tools/docker/linux_aarch64_native_image/Dockerfile:83]: It seems to be only needed for ancient versions of RH/CentOS e.g. `CentOS Linux release 7.1.1503 (Core)`, where certain macros defined  in `/usr/include/inttypes.h` are guarded by `_STDC_FORMAT_MACROS`. Hence its not strictly required for our docker images which are on a later version of the OS. I'll update the Docker files anyway, for consistency.

Comment (droberts195) [reply dev-tools/docker/linux_aarch64_native_image/Dockerfile:83]: So are you saying CentOS 7.8 has it but CentOS 7.1 doesn't? Both our Docker images are based on CentOS 7 latest patch.

It's fine to just include it in the commands then. Don't actually spend time rebuilding the images.

Comment (edsavage) [reply dev-tools/docker/linux_aarch64_native_image/Dockerfile:83]: Cool, yes I've found its only needed on our `marple` server. None of the docker images require it.

Comment (droberts195) [reply dev-tools/docker/linux_aarch64_native_image/Dockerfile:83]: In that case please add a TODO to the instructions to remove it once we no longer build on CentOS 7.

File: include/core/CMemoryDef.h
Comment (droberts195) [inline include/core/CMemoryDef.h:212]: You can do this without resorting to floating point functions like this:

```
((t.bucket_count() + 2 * sizeof(std::size_t) - 1) / 2 * sizeof(std::size_t)) * sizeof(std::size_t)
```

The compiler can simplify that better at compile time and avoid a function call at runtime.

Also, this formula could go in a new function `std::size_t bucketGroupOverhead(std::size_t bucketCount)` that could then have a Doxygen comment explaining what it's doing and linking to the Boost unordered documentation. That function should get inlined by the compiler if it's defined higher up the same file, so won't really add any overhead.

File: include/core/CMemoryDef.h
Comment (droberts195) [inline include/core/CMemoryDef.h:498]: Even though it's not a problem you introduced, I just noticed that this method is not using a helper function that the others are.

```suggestion
        (t.bucket_count() * memory::storageNodeOverhead(t)) +
        (t.size() * (sizeof(K) + sizeof(V) + memory::storageNodeOverhead(t))) +
```

File: lib/maths/analytics/unittest/CBoostedTreeTest.cc
Comment (droberts195) [inline lib/maths/analytics/unittest/CBoostedTreeTest.cc:3093]: This fudge factor shouldn't be in the test - it should be in the core code in `core::CDataFrame::estimateMemoryUsage`. Otherwise this test will be asserting that customers will have problems in production.

File: lib/core/CDataFrame.cc
Comment (droberts195) [inline lib/core/CDataFrame.cc:518]: Since this is only a crude estimate, is it really worth rounding to a multiple of `sizeof(std::size_t)`?

```suggestion
    std::size_t additionalMemory{
        static_cast<std::size_t>(estimatedMemoryUsage * boostUnorderedContainerMemorySlackPercent / 100)};
```

File: lib/core/CDataFrame.cc
Comment (droberts195) [inline lib/core/CDataFrame.cc:511]: I think there was also underestimation already in the `TStrVecVec` value, since the existing estimate assumes all the nested vectors are empty. It's probably worth making this fudge factor less specific to the Boost changes, and call the variable something more generic.

File: lib/api/CAnomalyJobConfig.cc
Comment (droberts195) [inline lib/api/CAnomalyJobConfig.cc:718]: Please give this method a more precise name. It doesn't appear to parse the full analysis config. Maybe something like `reparseDetectorsFromStoredConfig` (or something else that says what it does).

File: include/api/CAnomalyJobConfig.h
Comment (droberts195) [inline include/api/CAnomalyJobConfig.h:251]: Please add Doxygen comments to this method and the one immediately above saying why they do and why they're different.

**Review by droberts195 — APPROVED** [review]
LGTM if this passes CI

🤞

**Pull request merged**
